### PR TITLE
Emacs: Dark theme fixes and modernisation

### PR DESCRIPTION
**Pull request by Wilfred**
The Emacs mode in this repository has colors defined that don't work well in dark themes.

Before:

caml_before

After:

caml_after

This also affects packages like merlin, which use the styles from the caml-types.el file. In the screenshot above, I'm running the `merlin-type-enclosing` command.

I've also updated the face definitions to correspond to Emacs best practices. See the individual commit messages for more details.

(This is my first PR to ocaml itself, so please let me know if I've done anything silly!)

### CODE REVIEW

**Comment by pmetzger** [comment]
Does it still work well with light themes?

**Comment by gasche** [comment]
It would be nice to have some feedback from other people who know about the current good practices for Emacs configuration. For example (but other volunteers please feel free to chime in), @monnier, @cpitclaudel, would you by chance be able to comment?

**Comment by dbuenzli** [comment]
FWIW I'm using a dark theme and the only thing I had to change because of readability issues was:

```
(if window-system
    (progn
      (require 'caml-font)
      (set-face-foreground 'caml-font-doccomment-face "#cb4b16")))
```

but I guess the point of theming should be that you don't define the colors in the mode no ?

**Comment by cpitclaudel** [comment]
> For example (but other volunteers please feel free to chime in), @monnier, @cpitclaudel, would you by chance be able to comment?

These change look reasonable to me.  It's a mix of changing the code to use appropriate macros and adding a variant for dark background.

> you don't define the colors in the mode no ?

The mode should provide a default, and themes can override it :)

**Review by gasche — APPROVED** [review]
I'm approving based on @cpitclaudel's review (thanks!), but I'll still wait a few more days, in case people have objections, before merging.

**Comment by Wilfred** [comment]
> Does it still work well with light themes?

Yep, colours with light themes will be unchanged by this PR.

**Comment by Wilfred** [comment]
I see the CI is complaining that I haven't written a change entry. Should I write one for this editor tooling change? This change does not require users to do anything.

**Comment by gasche** [comment]
See our [CONTRIBUTING.md#Changes](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#changelog) guideline. This is your choice to make; there is no dumb contribution, so it would make sense for you to list your work in the Changes file and be credited for it. Also, it If you don't think it is worth the hassle, I can merge without a Changes entry -- I know that the rest of the CI passes, and we can tag this PR to exclude it from this test to get a green dot if we want.

**Comment by trefis** [comment]
I think a Changes entry is appropriate here.
@Wilfred: do you mind adding one now? I'll be happy to merge after.

**Comment by Wilfred** [comment]
@trefis okey-doke, I've updated `Changes`. Let me know if it needs fiddling in any way :)

**Comment by gasche** [comment]
The credits in parentheses should be on their own line, and it would be nice to credit your reviewer:

```
- GPR#1865: support dark themes in Emacs, and clean up usage of
  deprecated Emacs APIs
  (Wilfred Hughes, review by Clément Pit-Claudel)
```

**Comment by Wilfred** [comment]
Woops! Fixed.

**Pull request merged**
